modules/apps/27-interchain-accounts/types/packet.go

```diff
@@ -48,6 +54,109 @@ func (iapd InterchainAccountPacketData) GetBytes() []byte {
 	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(&iapd))
 }
 
+/*
+
+ADR-8 CallbackPacketData implementation
+
+InterchainAccountPacketData implements CallbackPacketData interface. This will allow middlewares targeting specific VMs
+to retrieve the desired callback address for the ICA packet on the source chain. Destination callback addresses are not
+supported for ICS 27.
+
+The Memo is used to set the desired callback addresses.
+
+The Memo format is defined like so:
+
+```json
+{
+	// ... other memo fields we don't care about
+	"callback": {
+		"src_callback_address": {contractAddrOnSourceChain},
+
+		// optional fields
+		"src_callback_msg": {jsonObjectForSourceChainCallback},
+	}
+}
+```
+
+*/
+
+// GetSourceCallbackAddress returns the source callback address provided in the packet data memo.
+// If no callback address is specified, an empty string is returned.
+//
+// The memo is expected to specify the callback address in the following format:
+// { "callback": { "src_callback_address": {contractAddrOnSourceChain}}
+//
+// ADR-8 middleware should callback on the returned address if it is a PacketActor
+// (i.e. smart contract that accepts IBC callbacks).
+func (iapd InterchainAccountPacketData) GetSourceCallbackAddress() string {
+	callbackData := iapd.GetCallbackData()
+	if callbackData == nil {
+		return ""
+	}
+
+	callbackAddr, ok := callbackData["src_callback_address"].(string)
+	if !ok {
+		return ""
+	}
+
+	return callbackAddr
+}
+
+// GetDestCallbackAddress returns an empty string. Destination callback addresses
+// are not supported for ICS 27. This feature is natively supported by
+// interchain accounts host submodule transaction execution.
+func (iapd InterchainAccountPacketData) GetDestCallbackAddress() string {
+	return ""
+}
+
+// GetUserDefinedCustomMessage returns the custom message provided in the packet data memo.
+// Custom message is expected to be base64 encoded.
+// If no custom message is specified, nil is returned.
+func (iapd InterchainAccountPacketData) GetUserDefinedCustomMessage() []byte {
+	callbackData := iapd.GetCallbackData()
+	if callbackData == nil {
+		return nil
+	}
+
+	callbackMsg, ok := callbackData["src_callback_msg"].(string)
+	if !ok {
+		return nil
+	}
+
+	// base64 decode the callback message
+	base64DecodedMsg, err := base64.StdEncoding.DecodeString(callbackMsg)
+	if err != nil {
+		return nil
+	}
+
+	return base64DecodedMsg
+}
+
+// UserDefinedGasLimit returns 0 (no-op). The gas limit of the executing
+// transaction will be used.
+func (iapd InterchainAccountPacketData) UserDefinedGasLimit() uint64 {
+	return 0
+}
+
+func (iapd InterchainAccountPacketData) GetCallbackData() map[string]interface{} {
```

Comment:
Can you add a docstring explaining this is used internally to satisfy the other interface functions. Does it even need to be exported?